### PR TITLE
[bitchute] extractor overhaul

### PR DESCRIPTION
**Pull request by flashdagger**
</details>

use convenient parsing functions from `utils` to simplify the code
for channels provide info of type `url_transparent` with additional properties, which allows for faster filtering/searching

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/bitchute.py [hunk 1]
```diff
@@ -69,24 +69,16 @@ def _real_extract(self, url):
...
+        # some videos have empty string as description, default suppresses the warning
+        description = self._og_search_description(webpage, default=None)
+        thumbnail = self._og_search_thumbnail(webpage)
+        uploader = clean_html(get_element_by_class('owner', webpage))
+        publish_date = clean_html(get_element_by_class('video-publish-date', webpage))
         upload_date = unified_strdate(self._search_regex(
-            r'class=["\']video-publish-date[^>]+>[^<]+ at \d+:\d+ UTC on (.+?)\.',
-            webpage, 'upload date', fatal=False))
+            r'at \d+:\d+ UTC on (.+?)\.', publish_date, 'upload date', fatal=False)
+        )
```
Comment (pukkandan) [inline yt_dlp/extractor/bitchute.py:81]:
```suggestion
            r'at \d+:\d+ UTC on (.+?)\.', publish_date, 'upload date', fatal=False))
```

File: yt_dlp/extractor/bitchute.py [hunk 2]
```diff
@@ -69,24 +69,16 @@ def _real_extract(self, url):
...
-            webpage, 'description', fatal=False)
-        thumbnail = self._og_search_thumbnail(
-            webpage, default=None) or self._html_search_meta(
-            'twitter:image:src', webpage, 'thumbnail')
-        uploader = self._html_search_regex(
-            (r'(?s)<div class=["\']channel-banner.*?<p\b[^>]+\bclass=["\']name[^>]+>(.+?)</p>',
-             r'(?s)<p\b[^>]+\bclass=["\']video-author[^>]+>(.+?)</p>'),
-            webpage, 'uploader', fatal=False)
-
+        # some videos have empty string as description, default suppresses the warning
```
Comment (pukkandan) [inline yt_dlp/extractor/bitchute.py:74]:
```suggestion
```

File: yt_dlp/extractor/bitchute.py [hunk 3]
```diff
@@ -42,12 +47,7 @@ def _real_extract(self, url):
                 'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/69.0.3497.57 Safari/537.36',
             })
 
-        title = self._html_search_regex(
-            (r'<[^>]+\bid=["\']video-title[^>]+>([^<]+)', r'<title>([^<]+)'),
-            webpage, 'title', default=None) or self._html_search_meta(
-            'description', webpage, 'title',
-            default=None) or self._og_search_description(webpage)
-
+        title = self._html_extract_title(webpage) or self._og_search_title(webpage)
```
Comment (pukkandan) [inline yt_dlp/extractor/bitchute.py:50]: This is not just a refactor, but change where title is extracted from. Is that intentional?

File: yt_dlp/extractor/bitchute.py [hunk 4]
```diff
@@ -18,7 +23,7 @@ class BitChuteIE(InfoExtractor):
         'url': 'https://www.bitchute.com/video/UGlrF9o9b-Q/',
         'md5': '7e427d7ed7af5a75b5855705ec750e2b',
         'info_dict': {
-            'id': 'szoMrox2JEI',
+            'id': 'UGlrF9o9b-Q',
```
Comment (pukkandan) [inline yt_dlp/extractor/bitchute.py:26]: video id should not change if it can be helped

File: yt_dlp/extractor/bitchute.py (hunk 1 above)
Comment (pukkandan) [inline yt_dlp/extractor/bitchute.py:81]: These are now simple enough to be inlined in the return

File: yt_dlp/extractor/bitchute.py [hunk 5]
```diff
@@ -100,51 +92,82 @@ def _real_extract(self, url):
...
-                    ie=BitChuteIE.ie_key(), video_id=video_id)
+                    ie=BitChuteIE.ie_key(), video_id=video_id, url_transparent=True,
+                    title=clean_html(get_element_by_class(class_name['title'], video_html)),
+                    description=clean_html(
+                        get_element_by_class(class_name['description'], video_html)
+                    ),
+                    duration=parse_duration(get_element_by_class('video-duration', video_html)),
+                    view_count=parse_count(clean_html(
+                        get_element_by_class('video-views', video_html))
+                    ),
```
Comment (pukkandan) [inline yt_dlp/extractor/bitchute.py:162]: coding convention
```suggestion
                    description=clean_html(get_element_by_class(class_name['description'], video_html)),
                    duration=parse_duration(get_element_by_class('video-duration', video_html)),
                    view_count=parse_count(clean_html(get_element_by_class('video-views', video_html))),
```

File: yt_dlp/extractor/bitchute.py [hunk 6]
```diff
@@ -100,51 +92,82 @@ def _real_extract(self, url):
...
+                ))
+            if not results:
+                break
+            offset += len(results)
+            for result in results:
+                yield result
```
Comment (pukkandan) [inline yt_dlp/extractor/bitchute.py:168]: see above comment about yield

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/bitchute.py (hunk 4 above)
Comment (pukkandan) [reply yt_dlp/extractor/bitchute.py:26]: nvm, I see this is a bug in the test defenition

File: yt_dlp/extractor/bitchute.py [hunk 7]
```diff
@@ -100,51 +92,82 @@ def _real_extract(self, url):
...
-                break
-            offset += len(video_ids)
-            for video_id in video_ids:
-                yield self.url_result(
+            results = []
+            class_name = self.HTML_CLASS_NAMES[playlist_type]
+            for video_html in get_elements_html_by_class(class_name['container'], html):
+                match = re.search(r'<a\b[^>]+\bhref=["\']/video/(?P<id>[^"\'/]+)', video_html)
+                if not match:
+                    continue
```
[the remark this answers is not in the document]
Comment (flashdagger) [reply yt_dlp/extractor/bitchute.py:150]: Old code was just looking for the video ids. Since the html provides more metadata (title, description, duration, view count) it can be extracted as well. When using e.g. `--match-filter "title ~= <keyword>"`  yt-dlp does not need to extract t every single video in the playlist which saves time since bitchute requests are not the fastest.

File: yt_dlp/extractor/bitchute.py (hunk 6 above)
Comment (flashdagger) [reply yt_dlp/extractor/bitchute.py]: please see f9346036175805ad69752aa8902c63d5f684d896

File: yt_dlp/extractor/bitchute.py (hunk 3 above)
Comment (flashdagger) [reply yt_dlp/extractor/bitchute.py]: Yes. The old implementation is 4 years old. Bitchute's HTML might change over time, but is equal for all videos.
And if there is a corner case one should provide a test for it.

File: yt_dlp/extractor/bitchute.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (pukkandan) [reply yt_dlp/extractor/bitchute.py]: I was asking about the specific code that I highlighted in my comment. The old code was extracting it with regex, and the new code with get_elements_html_by_class, but ends up giving more complicated code. So I was wondering if there was a reason you changed it

Anyway, never mind. It doesnt really matter

File: yt_dlp/extractor/bitchute.py [hunk 8]
```diff
@@ -69,82 +69,101 @@ def _real_extract(self, url):
...
+            if data.get('success') is not True:
+                return
+            class_name = self.HTML_CLASS_NAMES[playlist_type]
+            for video_html in get_elements_html_by_class(class_name['container'], data.get('html')):
+                match = re.search(r'<a\b[^>]+\bhref=["\']/video/(?P<id>[^"\'/]+)', video_html)
+                video_id = match and match.group('id')
                 yield self.url_result(
                     'https://www.bitchute.com/video/%s' % video_id,
-                    ie=BitChuteIE.ie_key(), video_id=video_id)
+                    ie=BitChuteIE, video_id=video_id, url_transparent=True,
```
Comment (pukkandan) [inline yt_dlp/extractor/bitchute.py:148]: Why transparent?

**Review by pukkandan — APPROVED** [review]

File: yt_dlp/extractor/bitchute.py (hunk 8 above)
Comment (flashdagger) [reply yt_dlp/extractor/bitchute.py:148]: Because `duration` and `view_count` cannot be extracted by `BitChuteIE`, so they will be propagated.
https://github.com/yt-dlp/yt-dlp/blob/1305b659ef2bf3c76851b9400c7ac4a8f100fce2/yt_dlp/YoutubeDL.py#L1678-L1680

Does this have any other implications?

**Comment by flashdagger** [comment]
I just noticed the filed issues complaining about missing Bitchute videos.
In that case we should keep the custom exception, to avoid unnecessary issues:
```
if not formats:
    self.raise_no_formats(
        'Video is unavailable. Please make sure this video is playable in the browser '
        'before reporting this issue.', expected=True, video_id=video_id)
```

**Comment by pukkandan** [comment]
Ideally, people would have the common sense to check if the video is playable before opening an issue. Or at the very least, do so when the issue template asks them to. But I digress. 

A custom error message does sound like a good idea

**Comment by flashdagger** [comment]
@pukkandan 
Since yt-dlp takes care of most web-related problems, the format extraction can be very simple.

The additional webseed URLs are nowhere to be found. There are even articles saying Bitchute is not a peer-to-peer network despite it claims.

So Bitchute actually provides one download URL per video. 
To check, if the video is actually downloadable users can provide optionally `--check-formats` option. Thus we need only one webpage request per video, which speeds up playlist processing and should always provide a valid info...

**Comment by pukkandan** [comment]
That last commit is very problematic. 

> remove format sorting

You must always do sorting when `formats` list is returned, even with only one entry

> remove code searching for webseed urls, Bitchute now only provides optional Magnet Links

Remove the webseed extraction only if you are 100% sure it doesnt work in any cases. If not, leave it in.

> don't check media availability, user can force it with '--check-formats'

Weren't you just saying that the error message should be improved? This makes it only worse. One additional head request is worth it here imo.

**Comment by flashdagger** [comment]
> That last commit is very problematic.
> 
> > remove format sorting
>
> You must always do sorting when `formats` list is returned, even with only one entry

Ok.
 
> > remove code searching for webseed urls, Bitchute now only provides optional Magnet Links
> 
> Remove the webseed extraction only if you are 100% sure it doesnt work in any cases. If not, leave it in.

There is never 100%. But how do you avoid code bloat? Here are my arguments:
* I checked the commits, the code comes from youtube-dl around 2019, there is no valid testcase
* When you research the topic on the web, you find information that BitChute started with the P2P approach, but in the end did not implement it
* now they only provide a Magnet link (which is not parsed)
* I yet didn't find a working webtorrent URL hosted on Bitchute, they all return 404
 
> > don't check media availability, user can force it with '--check-formats'
> 
> Weren't you just saying that the error message should be improved? This makes it only worse. One additional head request is worth it here imo.

I see. It all depends on the user. I prefer to get all the metadata, even if the video cannot be downloaded. But other people might be puzzled. Now I get the concrete HTTP error message, when I really want to download the video. This can have different reasons (404 Not found, connection refused, bad gateway etc.) so I get more context. 
Some videos can be even downloaded if I exchange the server URL (https://seed154.bitchute.com/1VBwRfyNcKdX/2s6B3nZjAk7R.mp4 gives 404, but https://seed150.bitchute.com/1VBwRfyNcKdX/2s6B3nZjAk7R.mp4 works)

The final decision is on you. Previous implementation was https://github.com/yt-dlp/yt-dlp/commit/3bc40db0b9454637eca33876cf3c2262c7d7eb8e

**Comment by pukkandan** [comment]
> I prefer to get all the metadata, even if the video cannot be downloaded. But other people might be puzzled. 

Due to legacy reasons, the default behavior has to be to raise error when there is only metadata without downloadable formats. But there's `--ignore-no-formats-error` to change that if u need to. This is the reason extractor code uses `self.raise_no_formats` instead of directly `raise ExtractorError`.

> Some videos can be even downloaded if I exchange the server URL ([seed154.bitchute.com/1VBwRfyNcKdX/2s6B3nZjAk7R.mp4](https://seed154.bitchute.com/1VBwRfyNcKdX/2s6B3nZjAk7R.mp4) gives 404, but [seed150.bitchute.com/1VBwRfyNcKdX/2s6B3nZjAk7R.mp4](https://seed150.bitchute.com/1VBwRfyNcKdX/2s6B3nZjAk7R.mp4) works)

Interesting, I agree this usecase cannot be covered by the older implementation. Though, I'm now thinking we could check formats only when this flag is not passed? That should address both usecases. Normally, I don't like when extractors directly access params, but it feels appropriate in this case. Let me know what u think

On a related note, how did you get the working URL? Just trial and error? Or is there some way to add that logic to the code?

**Comment by flashdagger** [comment]
The last implementation now checks all URLs with HEAD requests until `--no-check-formats` is specified. So the default behavior should be identical.

> On a related note, how did you get the working URL? Just trial and error? Or is there some way to add that logic to the code?

Lots of requests. Then I tried to ping the seedXXX servers. Here are the ones that are available from German location:
```
seed103.bitchute.com from seed103.bitchute.com (69.197.147.170)
seed104.bitchute.com from seed104.bitchute.com (204.12.250.83)
seed106.bitchute.com from seed106.bitchute.com (204.12.250.171)
seed108.bitchute.com from seed108.bitchute.com (204.12.250.84)
seed116.bitchute.com from seed116.bitchute.com (172.67.68.161)
seed117.bitchute.com from seed116.bitchute.com (172.67.68.161)
seed118.bitchute.com from seed118.bitchute.com (104.26.6.35)
seed119.bitchute.com from seed119.bitchute.com (104.26.7.35)
seed122.bitchute.com from seed122.bitchute.com (142.54.188.114)
seed124.bitchute.com from seed124.bitchute.com (145.239.254.226)
seed126.bitchute.com from seed126.bitchute.com (192.187.114.98)
seed128.bitchute.com from seed128.bitchute.com (69.30.243.138)
seed129.bitchute.com from seed129.bitchute.com (209.141.60.40)
seed130.bitchute.com from seed130.bitchute.com (198.98.54.246)
seed131.bitchute.com from seed129.bitchute.com (209.141.60.40)
seed132.bitchute.com from seed132.bitchute.com (173.208.176.130)
seed150.bitchute.com from seed150.bitchute.com (69.30.230.34)
seed151.bitchute.com from ip220.ip-51-77-124.eu (51.77.124.220)
seed152.bitchute.com from mexico-gob.com (69.30.224.114)
seed153.bitchute.com from kansas.internetdeoccidente.net (199.168.96.26)
seed154.bitchute.com from seed154.bitchute.com (107.150.55.234)
seed160.bitchute.com from seed116.bitchute.com (172.67.68.161)
seed167.bitchute.com from seed167.bitchute.com (142.54.189.194)
seed171.bitchute.com from seed171.bitchute.com (173.208.203.250)
seed177.bitchute.com from seed177.bitchute.com (173.208.246.162)
seed200.bitchute.com from seed200.bitchute.com (185.238.113.1)
seed302.bitchute.com from seed302.bitchute.com (204.12.211.194)
seed303.bitchute.com from seed303.bitchute.com (204.12.207.58)
seed305.bitchute.com from seed305.bitchute.com (192.187.97.234)
seed306.bitchute.com from seed306.bitchute.com (69.197.138.26)
seed307.bitchute.com from seed307.bitchute.com (204.12.220.234)
```

seed150...seed154 seemed to be the most promising candidates. Try for older entries of bigger channels and you'll see the success:
```
yt-dlp https://www.bitchute.com/channel/sanity-for-sweden/ --print "%(playlist_index)5s/%(playlist_count)s %(id)s %(urls)s" --playlist-items 1600:
```

Of course this can vary from location. But you can only gain by testing some more requests...

File: yt_dlp/extractor/bitchute.py [hunk 9]
```diff
@@ -109,51 +113,106 @@ def _real_extract(self, url):
...
+                    'duration': 16,
+                    'view_count': int,
+                },
+            }
+        ],
+        'params': {
+            'skip_download': True,
+            'check_formats': False,
+            'match_filter': match_filter_func('id=UGlrF9o9b-Q'),
+        },
```
Comment (pukkandan) [inline yt_dlp/extractor/bitchute.py:146]: Why is the test defined like this?

File: yt_dlp/extractor/bitchute.py [hunk 10]
```diff
@@ -109,51 +113,106 @@ def _real_extract(self, url):
...
+            if data.get('success') is not True:
+                return
+            class_name = self.HTML_CLASS_NAMES[playlist_type]
+            for video_html in get_elements_html_by_class(class_name['container'], data.get('html')):
+                match = re.search(r'<a\b[^>]+\bhref=["\']/video/(?P<id>[^"\'/]+)', video_html)
+                video_id = match and match.group('id')
                 yield self.url_result(
-                    'https://www.bitchute.com/video/%s' % video_id,
-                    ie=BitChuteIE.ie_key(), video_id=video_id)
+                    f'https://www.bitchute.com/video/{video_id}',
```
Comment (pukkandan) [inline yt_dlp/extractor/bitchute.py]: This gives `/video/None` in the second test case

File: yt_dlp/extractor/bitchute.py [hunk 11]
```diff
@@ -109,51 +113,106 @@ def _real_extract(self, url):
...
-        channel_id = self._match_id(url)
+        playlist_type, playlist_id = self._match_valid_url(url).group('type', 'id')
+
+        webpage = self._download_webpage(
+            f'https://www.bitchute.com/{playlist_type}/{playlist_id}/', video_id=playlist_id)
+        title = self._html_extract_title(webpage, default=None)
+        description = self._html_search_meta(
+            ('description', 'og:description', 'twitter:description'), webpage, default=None)
+        playlist_count = int_or_none(self._html_search_regex(
+            r'<span>(\d+) +videos?</span>', webpage, 'playlist count', default=None), default='N/A')
```
Comment (pukkandan) [inline yt_dlp/extractor/bitchute.py]: defaut=None

File: yt_dlp/extractor/bitchute.py [hunk 12]
```diff
@@ -109,51 +113,106 @@ def _real_extract(self, url):
...
+    }
 
-    def _entries(self, channel_id):
-        channel_url = 'https://www.bitchute.com/channel/%s/' % channel_id
-        offset = 0
-        for page_num in itertools.count(1):
+    def _entries(self, playlist_type, playlist_id):
+        playlist_url = f'https://www.bitchute.com/{playlist_type}/{playlist_id}/'
+
+        def fetch_entries(page_num):
```
Comment (pukkandan) [inline yt_dlp/extractor/bitchute.py]: closure is unnecessary

File: yt_dlp/extractor/bitchute.py [hunk 13]
```diff
@@ -109,51 +113,106 @@ def _real_extract(self, url):
...
-            video_ids = re.findall(
-                r'class=["\']channel-videos-image-container[^>]+>\s*<a\b[^>]+\bhref=["\']/video/([^"\'/]+)',
-                html)
-            if not video_ids:
-                break
-            offset += len(video_ids)
-            for video_id in video_ids:
+            if data.get('success') is not True:
+                return
+            class_name = self.HTML_CLASS_NAMES[playlist_type]
```
Comment (pukkandan) [inline yt_dlp/extractor/bitchute.py]: confusing var name for a dict

**Review by pukkandan — CHANGES_REQUESTED** [review]
This PR started out improving playlists, but you ended up also working on BitchuteIE too. Pls make separate PRs in future for such unrelated changes

File: yt_dlp/extractor/bitchute.py (hunk 9 above)
Comment (flashdagger) [reply yt_dlp/extractor/bitchute.py:146]: When using a channel or playlist url we get more fields in the metadata. So I wanted to test the same video as in BitChuteIE.

**Comment by flashdagger** [comment]
> This PR started out improving playlists, but you ended up also working on BitchuteIE too. Pls make separate PRs in future for such unrelated changes

Sometimes I lose track and see everything interrelated. But since I package it into separate commits I can always move changes to another PR. No problem for me.

**Comment by pukkandan** [comment]
> > This PR started out improving playlists, but you ended up also working on BitchuteIE too. Pls make separate PRs in future for such unrelated changes
> 
> @pukkandan Should I open a new PR, so we can continue with changes on `BitChuteChannelIE`?

No need. Since I already merged the rest, this is fine
